### PR TITLE
features and fixes for v1.1.1 release

### PR DESCRIPTION
**Pull request by akornich**
Addressed work items: 
#103 #104 #105 #106 #107 #111 #112

### CODE REVIEW

File: Rollbar/Common/ReconfigurableBase.cs [hunk 1]
```diff
@@ -2,12 +2,14 @@
 {
     using System;
     using Rollbar.Utils;
+    using Xamarin.iOS.Foundation;
```
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Common/ReconfigurableBase.cs]: nope, this is just a namespace our custom PreserveAttribute type is expected to live in. no references to Xamarin libs at all.

File: Rollbar/Common/StringUtility.cs [hunk 2]
```diff
@@ -27,5 +27,75 @@ public static string Combine(IEnumerable<string> substrings, string substringSep
...
+            }
+
+            byte[] inputEncodedBytes = encoding.GetBytes(input);
+            Assumption.AssertEqual(inputEncodedBytes.Length, encoding.GetByteCount(input), nameof(inputEncodedBytes.Length));
+            if (inputEncodedBytes.Length <= encodedBytesLimit)
+            {
+                return input; //nothing to truncate...
+            }
+
+            int truncaionIndicatorTotalByte = encoding.GetByteCount(truncaionIndicator);
```
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Common/StringUtility.cs]: oops. fixed. thanks!

File: Rollbar/Common/StringUtility.cs [hunk 3]
```diff
@@ -27,5 +27,75 @@ public static string Combine(IEnumerable<string> substrings, string substringSep
...
+            int totalBytes = encodedBytesLimit - truncaionIndicatorTotalByte;
+            if (totalBytes < 0)
+            {
+                totalBytes = 0;
+            }
+            string truncatedInput = encoding.GetString(inputEncodedBytes, 0, totalBytes);
+            if ((totalBytes > 0) 
+                && (truncatedInput[truncatedInput.Length - 1] != input[truncatedInput.Length - 1])
```
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Common/StringUtility.cs]: this line 
string truncatedInput = encoding.GetString(inputEncodedBytes, 0, totalBytes);
produces truncatedInput string based on the first totalBytes out of the initial non-truncated inputEncodedBytes. GetString(...) method does its best at decoding the subset of bytes. Depending where exactly cut-off point of the bytes falls, we can have either correct or incorrect decoding of the last char. For example, GetString(...) can insert unknown char symbol if it had hard time interpreting the incomplete byte chunk for the last char of the string. So, here we essentially cleaning up any such garbage when applicable.

File: Rollbar/Common/StringUtility.cs [hunk 4]
```diff
@@ -27,5 +27,75 @@ public static string Combine(IEnumerable<string> substrings, string substringSep
...
+                )
+            {
+                truncatedInput = truncatedInput.Substring(0, truncatedInput.Length - 1);
```
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Common/StringUtility.cs]: in the call
truncatedInput = truncatedInput.Substring(0, truncatedInput.Length - 1);
the second argument of the Substring(...) method represents length of the substring (not the last index of the char for substring).

File: Rollbar/Diagnostics/Assumption.cs [hunk 5]
```diff
@@ -374,7 +374,7 @@ public static DateTimeOffset AssertGreaterThanOrEqual(DateTimeOffset value, Date
         /// <returns></returns>
         public static int AssertLessThan(int value, int expectedValue, string parameterName)
         {
-            if (value < expectedValue)
```
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Diagnostics/Assumption.cs]: before, only "Greater" alternatives were used. Now, i happened to use the "Less..." first time and noticed the typo in multiple variants of it. If i remember correctly, i implemented the "Greater" ones first and cloned and modified (as it appears now, not completely) into "Less" ones.

File: Rollbar/RollbarClient.cs [hunk 6]
```diff
@@ -41,10 +44,20 @@ public RollbarResponse PostAsJson(Payload payload, IEnumerable<string> scrubFiel
...
 
+        private readonly IterativeTruncationStrategy _payloadTruncationStrategy = new IterativeTruncationStrategy();
+
         public async Task<RollbarResponse> PostAsJsonAsync(Payload payload, IEnumerable<string> scrubFields)
         {
             Assumption.AssertNotNull(payload, nameof(payload));
 
+            if (this._payloadTruncationStrategy.Truncate(payload) > this._payloadTruncationStrategy.MaxPayloadSizeInBytes)
+            {
+                throw new ArgumentOutOfRangeException(
```
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/RollbarClient.cs]: Good point! However, in RollbarQueueController class, on line 259 every PostAsJason(...) call is already in try-catch block and reports any caught exceptions as Rollbar internal events that the customers can tap into. This way we are safe, informative and not even attempting to send the over-sized payload.

File: Rollbar/Common/StringUtility.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rokob) [reply Rollbar/Common/StringUtility.cs]: Ah so unicode issues based on byte boundaries being different than character boundaries. Gotcha.

File: Rollbar/Common/StringUtility.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rokob) [reply Rollbar/Common/StringUtility.cs]: Well looks good to me once you push up the fix for this typo.

[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Common/StringUtility.cs]: ok, it's in...

File: Rollbar/Common/StringUtility.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (akornich) [reply Rollbar/Common/StringUtility.cs]: correct...

**Pull request merged**